subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java

```diff
@@ -54,28 +55,33 @@ public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionCon
         FileCollection sourceFiles = task.getInputs().getSourceFiles();
         if (task.getInputs().getHasSourceFiles() && sourceFiles.isEmpty()) {
             TaskArtifactState taskArtifactState = context.getTaskArtifactState();
-            Set<File> outputFiles = taskArtifactState.getExecutionHistory().getPreviousOutputs();
+            TaskExecutionHistory executionHistory = taskArtifactState.getExecutionHistory();
+            Set<File> outputFiles = executionHistory.getPreviousOutputs();
             if (outputFiles == null) {
                 state.setOutcome(TaskExecutionOutcome.NO_SOURCE);
                 LOGGER.info("Skipping {} as it has no source files and no history of previous output files.", task);
             } else if (outputFiles.isEmpty()) {
                 state.setOutcome(TaskExecutionOutcome.NO_SOURCE);
                 LOGGER.info("Skipping {} as it has no source files and no previous output files.", task);
             } else {
+                boolean cleanupDirectories = true;
+                if (executionHistory.getOverlappingOutputs() != null) {
+                    LOGGER.info("No leftover directories for {} will be deleted since overlapping outputs where detected.", task);
+                    cleanupDirectories = false;
+                }
```

Comment:
Pet peeve, but this could be done in a way that the value of `cleanupDirectories` is only set once and not changed later. I think that's a better for both human and machine.